soroban-env-common/src/symbol.rs

```diff
@@ -224,40 +228,38 @@ impl<const N: u32> TryFrom<&StringM<N>> for SymbolSmall {
 }
 
 impl SymbolSmall {
-    #[doc(hidden)]
-    pub const fn validate_char(ch: char) -> Result<(), SymbolError> {
-        match SymbolSmall::encode_char(ch) {
+    const fn validate_byte(b: u8) -> Result<(), SymbolError> {
+        match Self::encode_byte(b) {
             Ok(_) => Ok(()),
             Err(e) => Err(e),
         }
     }
 
-    const fn encode_char(ch: char) -> Result<u64, SymbolError> {
-        let v = match ch {
-            '_' => 1,
-            '0'..='9' => 2 + ((ch as u64) - ('0' as u64)),
-            'A'..='Z' => 12 + ((ch as u64) - ('A' as u64)),
-            'a'..='z' => 38 + ((ch as u64) - ('a' as u64)),
-            _ => return Err(SymbolError::BadChar(ch)),
+    const fn encode_byte(b: u8) -> Result<u8, SymbolError> {
+        let v = match b {
+            b'_' => 1,
+            b'0'..=b'9' => 2 + (b - b'0'),
+            b'A'..=b'Z' => 12 + (b - b'A'),
+            b'a'..=b'z' => 38 + (b - b'a'),
+            _ => return Err(SymbolError::BadChar(b)),
         };
         Ok(v)
     }
 
-    pub const fn try_from_bytes(b: &[u8]) -> Result<SymbolSmall, SymbolError> {
+    pub const fn try_from_bytes(bytes: &[u8]) -> Result<SymbolSmall, SymbolError> {
+        if bytes.len() > MAX_SMALL_CHARS {
+            return Err(SymbolError::TooLong(bytes.len()));
+        }
         let mut n = 0;
         let mut accum: u64 = 0;
-        while n < b.len() {
-            let ch = b[n] as char;
-            if n >= MAX_SMALL_CHARS {
-                return Err(SymbolError::TooLong(b.len()));
-            }
-            n += 1;
-            accum <<= CODE_BITS;
-            let v = match SymbolSmall::encode_char(ch) {
+        while n < bytes.len() {
```

Comment:
There's a behavior change in here that seems unnecessary. The length check is moved to be earlier so there are failure cases that might fail with a length check that would have previously failed with some other error if there were invalid bytes prior to the maximum length being reached. Is this behavior check unnecessary? If no, I think we could be surgical with this change so it's purely non-functional / refactor and has no impact on observable behavior where it doesn't need to.

There also seem to be other small changes to this function that also seem unnecessary, but probably have no functional impact, but it'd be easier to review this change if it was as minimal/surgical as possible. If the changes are required, no worries, it's just not obvious to me that they are.

Reply:
This doesn't change the observable behavior in host because it doesn't distinguish between symbol errors. I could revert this, but given that here we're changing the error enum itself (which is observable by a theoretical env-common user, which I'm not sure if exists), I thought we might as well make this function a bit more clear and have consistent symbol validation logic between short and long symbols (so that we always check len first, then characters).